File: Visitor/C++/Renderer.cpp
#include <iostream>

#include "Renderer.hpp"

#include "DrawableShape.hpp"
#include "Point.hpp"
#include "Line.hpp"

void Renderer::addShape(const DrawableShape& shape) {
    shapes.push_back(shape);
}

void Renderer::drawAll() const {
    for (auto& shape : shapes) {
        shape.get().draw(*this);
    }
}

void Renderer::draw(const Point& point) const {
    std::cout << "Drawing point " << point << std::endl;
}

void Renderer::draw(const Line& line) const {
    std::cout << "Drawing line " << line << std::endl;
}

File: Visitor/TypeScript/Renderer.ts
/* created by Ghabriel Nunes <ghabriel.nunes@gmail.com> [2017] */

import {DrawableShape} from "./DrawableShape"
import {Point} from "./Point"
import {Line} from "./Line"

export class Renderer {
	public addShape(shape: DrawableShape): void {
		this.shapes.push(shape);
	}

	public drawAll(): void {
		for (let shape of this.shapes) {
			shape.draw(this);
		}
	}

	public drawPoint(point: Point): void {
		console.log("Drawing point " + point.toString());
	}

	public drawLine(line: Line): void {
		console.log("Drawing line " + line.toString());
	}

	private shapes: DrawableShape[] = [];
}
